scarlet/source.py

```diff
@@ -3,344 +3,182 @@
-    if Ny<=1:
-        _Ny = img.shape[1]
-        Ny = 2*img.shape[1]-1
-        _img = np.zeros((img.shape[0], Ny, img.shape[2]))
-        _img[:,:_Ny-1] = np.fliplr(img[:,1:])
-        _img[:,_Ny-1:] = img[:]
-        _y = _Ny - 1
-    else:
-        _img = img.copy()
-    if Nx<=1:
-        _Nx = img.shape[2]
-        Nx = 2*img.shape[2]-1
-        __img = _img.copy()
-        _img = np.zeros((_img.shape[0], _img.shape[1], Nx))
-        _img[:,:,:_Nx-1] = np.fliplr(__img[:,:,1:])
-        _img[:,:,_Nx-1:] = __img
-        _x = _Nx - 1
-    cx, cy = Nx // 2, Ny // 2
-
-    morph = np.zeros((Ny,Nx))
-    # use the band with maximum flux for the source
-    band = np.argmax(_img[:,_y,_x])
-    morph[:] = _img[band,_y-cy:_y+cy+1,_x-cx:_x+cx+1]/source.sed[0,band]
-    morph = morph.reshape((morph.size,))
-    morph[morph<0] = 0
-    # Apply the appropriate constraints
-    if symmetric:
-        # Make the model symmetric
-        symmetric = morph[::-1]
-        morph = np.min([morph, symmetric], axis=0)
-    if monotonic:
-        # Make the model monotonic
-        prox_monotonic = operators.prox_strict_monotonic((Ny, Nx), thresh=0, use_nearest=False)
-        morph = prox_monotonic(morph.reshape(morph.size,), 0)
-    # Trim the source to set the new size
-    morph = morph.reshape(Ny,Nx)
-    
-    cutoff = blend._bg_rms[band]*thresh
-    cuts = morph>cutoff
-    # Make sure that the source has at least one source
-    # above the cutoff value
-    if np.sum(cuts)==0:
-        msg = "Source centered at y={0}, x={1} has no flux above the cutoff ({2})"
-        raise SourceInitError(msg.format(_y, _x, cutoff))
-
-    ypix, xpix = np.where(cuts)
-    Ny = np.max(ypix)-np.min(ypix)
-    Nx = np.max(xpix)-np.min(xpix)
-    Ny += 1 - Ny % 2
-    Nx += 1 - Nx % 2
-    _cx = Nx//2
-    _cy = Ny//2
-    morph = morph[cy-_cy:cy+_cy+1, cx-_cx:cx+_cx+1]
-    source.resize([Ny, Nx])
-    # Make the bulge size smaller
-    # TODO: improve this algorithm
-    x = np.arange(Nx)
-    y = np.arange(Ny)
-    X,Y = np.meshgrid(x,y)
-    X = X - _cx
-    Y = Y - _cy
-    distance = np.sqrt(X**2+Y**2)
-    _morph = np.zeros_like(morph)
-    cut = distance<min(Ny*disk_ratio/2,Nx*disk_ratio/2)
-    _morph[cut] = morph[cut]*2/3
-    source.morph[0] = _morph.reshape((1,_morph.size))
-    _morph = morph-_morph
-    if monotonic:
-        prox_monotonic = operators.prox_strict_monotonic((Ny, Nx), thresh=0, use_nearest=False)
-        _morph = prox_monotonic(_morph.reshape(_morph.size,), 0)
-    source.morph[1] = _morph.reshape((1,_morph.size))
 
 class Source(object):
+
     """A single source in a blend
     """
-    def __init__(self, center, shape, K=1, psf=None, constraints=None, fix_sed=False, fix_morph=False,
-                 fix_frame=False, shift_center=0.2, init_func=init_templates):
+    def __init__(self, sed, morph_image, constraints=None, center=None, psf=None, fix_sed=False, fix_morph=False,
+                 fix_frame=False, shift_center=0.2):
```

Comment:
Feel free to propose a better one. But I think fundamentally a source is given by sed and morphology, and the new API reflect that. It's then the duty of the derived class to determine whatever they initial sed and morphology ought to be, in correspondence with the image and the constraints.

One could argue, that the base class `Source` may not even need `constraints` because of what I wrote above, but I didn't find an elegant way to excise it.

Reply:
I'm torn. I agree that it should be perfectly valid to initialize a `Source` with a morphology and SED instead of using an internal function, so your initialization of the base class seems appropriate. However, it doesn't feel "pythonic" to create an inherited class that is only used for initialization. I think I would still prefer passing an initialization method to the `Source` class, so that it can be initialized with _either_ an SED *and* morphology, _or_ a function that accepts a `Source` as the first parameter to initialize the source. 

For example you could change the function signature and beginning of `Source.__init__` to:
```
def __init__(self, sed=None, morph_image=None, init_func=None, constraints=None, center=None, psf=None,
             fix_sed=False, fix_morph=False, fix_frame=False, shift_center=0.2):
    if sed is None:
        if morph_image != None:
            init_msg = "Source must be initialized with either an 'sed' and 'morph_image' or 'init_func'"
            raise SourceInitError(init_msg)
        sed, morph = init_func(self)
```

Initializing a new source would then be done with
```
init_func = partial(init_above_noise, img=img, bg_rms=bg_rms)
src = Source(init_func=init_func, center=center, ...)
```

Reply:
It's perfectly pythonic to do so. In some cases the derived Source classes will implement more than just an init.

However, I dislike the flexibility you want to include in the base class. Here's the problem I tried to solve: right now (and in your proposal here) it's *very* easy to create a Source for which the parameters of `__init__` are inconsistent or at least don't work well together.

Example: peak initialization with multiple source components makes no sense and will not work as intended. There are numerous cases I could think of, so why bother allowing them?

It's now up to the user to create a derived class that yields a consistent combination of sed, morphology, and constraints. You can't have that if the init function callback is available for the base class.

Reply:
Plus, the only inelegance of the setup arisen in cases when the initialization function (now called in the derived class) needs to talk to the base class, e.g. for resizing the source box. If the init function (not `__init__`) lives outside of any class, one needs to pass `source` as first parameter. Moving those functions to class methods would solve that.